File: r/multiple_linear_regression.R
# ==========================================================
# 🧠 Multiple Linear Regression: 50 Startups Dataset (R)
# Author: Chukwuka Chijioke Jerry
# ==========================================================

# 📦 Load Required Libraries
library(caTools)
library(ggplot2)

# 📁 1. Import Dataset
dataset <- read.csv("../data/50_startups.csv")

# 🔄 2. Encode Categorical Data
dataset$State <- factor(dataset$State,
                        levels = unique(dataset$State),
                        labels = 1:length(unique(dataset$State)))

# 🔀 3. Split Dataset into Training and Test Sets
set.seed(123)
split <- sample.split(dataset$Profit, SplitRatio = 0.8)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# 🧠 4. Fit Multiple Linear Regression Model
model <- lm(Profit ~ ., data = training_set)
summary(model)

# 📈 5. Predict on Test Set
predictions <- predict(model, newdata = test_set)

# 💾 6. Save Predictions to CSV
output <- data.frame(
  R.D.Spend = test_set$R.D.Spend,
  ActualProfit = test_set$Profit,
  PredictedProfit = predictions
)

if (!dir.exists("../outputs")) dir.create("../outputs")
write.csv(output, "../outputs/MLR_predicted_profits_r.csv", row.names = FALSE)

# 🖼️ 7. Plot Actual vs Predicted
if (!dir.exists("../plots")) dir.create("../plots")
plot <- ggplot(output, aes(x = ActualProfit, y = PredictedProfit)) +
  geom_point(color = "blue", size = 3) +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") +
  xlab("Actual Profit") +
  ylab("Predicted Profit") +
  ggtitle("Actual vs Predicted Profit (R)") +
  theme_minimal()

ggsave("../plots/mlr_actual_vs_predicted_r.png", plot, width = 8, height = 5)

# 🔄 8. Backward Elimination
cat("\n🔄 Backward Elimination Process:\n")

reg1 <- lm(Profit ~ R.D.Spend + Administration + Marketing.Spend + State, data = dataset)
summary(reg1)

reg2 <- lm(Profit ~ R.D.Spend + Administration + Marketing.Spend, data = dataset)
summary(reg2)

reg3 <- lm(Profit ~ R.D.Spend + Marketing.Spend, data = dataset)
summary(reg3)

reg4 <- lm(Profit ~ R.D.Spend, data = dataset)
summary(reg4)

# ✅ Conclusion
cat("\n✅ Based on adjusted R², the model with 'R.D.Spend + Marketing.Spend' performs best.\n")

File: python/multiple_linear_regression.py
# multiple_linear_regression.py

"""
🏢 Multiple Linear Regression on 50 Startups Dataset

Predict company profit based on R&D Spend, Administration, Marketing Spend, and State.
Author: Chukwuka Chijioke Jerry
"""

# ===============================
# 📦 Import Libraries
# ===============================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# ===============================
# 📥 Load Dataset
# ===============================
dataset = pd.read_csv('../data/50_startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# ===============================
# 🔄 Encode Categorical Data
# ===============================
ct = ColumnTransformer(
    transformers=[('encoder', OneHotEncoder(drop='first'), [3])],
    remainder='passthrough'
)
X = ct.fit_transform(X)

# ===============================
# 🔀 Split Dataset
# ===============================
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# ===============================
# 🧠 Train the Model
# ===============================
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# ===============================
# 📊 Make Predictions
# ===============================
y_pred = regressor.predict(X_test)
results = pd.DataFrame({
    'Actual Profit': y_test,
    'Predicted Profit': y_pred
})

# ===============================
# 💾 Save Outputs
# ===============================
os.makedirs('../outputs', exist_ok=True)
results.to_csv('../outputs/MLR_predicted_profits_py.csv', index=False)

# ===============================
# 📈 Visualization
# ===============================
os.makedirs('../plots', exist_ok=True)

plt.figure(figsize=(8, 5))
sns.scatterplot(x=y_test, y=y_pred)
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], color='red', linestyle='--')
plt.xlabel('Actual Profit')
plt.ylabel('Predicted Profit')
plt.title('Actual vs Predicted Profit')
plt.grid(True)
plt.tight_layout()
plt.savefig('../plots/mlr_actual_vs_predicted_py.png')
plt.show()

# ===============================
# 📌 Model Coefficients
# ===============================
print("Intercept:", regressor.intercept_)
print("Coefficients:", regressor.coef_)

# ✅ Done
print("✅ Regression complete. Outputs saved in 'outputs/', plot saved in 'plots/'.")
